### DISCUSSION TITLE
Implement Another COM object

### DISCUSSION THREAD

**Discussion by kevinxxq** [halildurmus/win32#317] [Q&A] [answered]
Hi Tim,
Thank you for the package.

I would like to implement a third party COM object. I have checked the below docs:

- https://github.com/timsneath/win32/blob/main/CONTRIBUTING.md
- https://pub.dev/documentation/win32/latest/topics/com-topic.html and 
- https://github.com/timsneath/win32/tree/main/tool

but now I still don't have any idea how to do, as I understand:
1. If I would like to do that manually, I need to implement a class IAnotherCOM extends IUnknown, I need provide the CLSID_AnotherCOM and IID_AnotherCOM. And then I need cast the required functions manually.

2. If I would like to do that by your tools. I need get metadata just as _The metadata is in the form of an ECMA-335 specification file, generated from this repo: https://github.com/microsoft/win32metadata_ and add the interface name(as 'Windows.Win32.Globalization.ISpellChecker',) to https://github.com/timsneath/win32/blob/main/tool/metadata/generate_com_apis.dart then run generate.cmd

Please let me know if there is something wrong.
My question is that if I cast it manually how can I know the index of element? if I do that by tools, I can't find the interface name.

Thank you.

**Comment by kevinxxq** [comment]
After days' search, I need update the question:
1. From the https://docs.microsoft.com/en-us/dotnet/framework/tools/tlbimp-exe-type-library-importer. The output of Tlbimp.exe is a binary file (an assembly) that contains runtime metadata for the types defined within the original type library. 

I checked the file generated by Tlbimp.exe by ILSpy and there are information as metadata:
![image](https://user-images.githubusercontent.com/55800697/137757622-12f13b60-d7a2-461f-9a1c-4087d9cb649f.png)

2. As this post indicates, https://stackoverflow.com/questions/54375771/how-to-read-a-winmd-winrt-metadata-file
we can read and parse metadata as .winmd

Can we parse the metadata and make a projection between COM and dart class?

Thank you.

**Comment by timsneath** [comment]
You should be able to parse the metadata with https://pub.dev/packages/winmd. But you don't necessarily need metadata to build the interop class: you just need to understand the COM object you're wrapping. If you look at a class like https://github.com/timsneath/win32/blob/main/lib/src/com/ISpellCheckerFactory.dart, you'll see how to manually create a similar definition. 

A vtable is just a table of function pointers. If an interface inherits from `IUnknown`, then the vtable's first three entries point to the methods in `IUnknown`, and the fourth entry (index 3) will point to the first method in your interface. 

The existing projection tooling was only written with Win32 APIs in mind, unfortunately. I haven't tested it with other .winmd files.

**Comment by kevinxxq** [comment]
Thank you Tim,
Yes, I figured out the projection these days and my mistake is that the interface extends from IDisplatch, so the element should start from 7.

The other hand is that the .Net interop handles the COM very easy and no such  difficult things(only for me), I need count the elements and check every parameter.

For a specific question, do you know how to handle the below projection:
I have method as below

 [id(0x68030001), propget, helpstring("Get/assign wave,n,k values")]
        VARIANT nkTable();
        [id(0x68030001), propput, helpstring("Get/assign wave,n,k values")]
        void nkTable([in, out] VARIANT* rhs);

In .Net, it projection as below:
        [DispId(1745027073)]
        public virtual object get_nkTable();
        [DispId(1745027073)]
        public virtual void set_nkTable(ref object value);

In VBA it need ReDim xIndx(10, 3)

I don't know what is VARIANT*, I searched it and get a page https://docs.microsoft.com/en-us/office/vba/language/reference/user-interface-help/variant-data-type, it looks like a object(can be anything).

Please help me check how to projection it to dart types.

**Comment by timsneath** [comment]
The Win32 package has a `VARIANT` type that you can use here (or `Pointer<VARIANT>` for the `rhs` variable). 

https://pub.dev/documentation/win32/latest/win32/VARIANT-class.html

Variant is weakly typed and can contain almost any value; it's your responsibility to know which type to extract from it. For example, you can call `final dartValue = myVariant.uintVal` to get an unsigned 32-bit integer into `dartValue`.

**Comment by kevinxxq** [comment]
Thank you, Tim, Sorry for the stupid headache question wastes your time.
My data type is a 2 dimention array of float.

I tried to define a ARRAY.multi (https://api.dart.dev/stable/2.14.1/dart-ffi/Array/Array.multi.html) and then send to the function but failed.

Array<Array<Float>> array = Array.multi([4, 4]);

    return ptr.ref.lpVtbl.value
        .elementAt(22)
        .cast<Pointer<NativeFunction<_set_nkTable_Native>>>()
        .value
        .asFunction<_set_nkTable_Dart>()(
      ptr.ref.lpVtbl,
      array,
    );

if I define the function parameter as VARIANT as below:
typedef _set_nkTable_Native = Int32 Function(    Pointer obj, VARIANT indexArray);
typedef _set_nkTable_Dart = int Function(    Pointer obj, VARIANT indexArray);

the error is:
The argument type 'Array<NativeType>' can't be assigned to the parameter type 'VARIANT'.

if I define the function parameter as Array<Array<Float>> as below:
typedef _set_nkTable_Native = Int32 Function(
    Pointer obj, Array<Array<Float>> indexArray);
typedef _set_nkTable_Dart = int Function(
    Pointer obj, Array<Array<Float>> indexArray);

The error is:
The type argument for the pointer 'NativeFunction<Int32 Function(Pointer<NativeType>, Array<Array<Float>>)>' must be a 'NativeFunction' in order to use 'asFunction'.
Try changing the type argument to be a 'NativeFunction'.dart(non_native_function_type_argument_to_pointer)


Would you please help check this?

**Comment by timsneath** [comment, marked as answer]
We're some way from the win32 package at this point, but the main challenge you have is how to express an array in the `VARIANT` data type. 

The Dart definition of `VARIANT` is here:
https://github.com/timsneath/win32/blob/83e32cb0426b3ee066e1c5ea36d5e8d1b0756f7f/lib/src/structs.dart#L328

You need to set the `vt` to the right array type (Win32 has multiple), and then set `llVal` to the memory address. Beyond that, I can't help you, I'm afraid -- we're getting into the vagaries of calling `IDispatch` and OLE Automation from native code, rather than Dart interop.

**Comment by kevinxxq** [comment]
OK thank you.
